pkg/apb/dockerhub_registry.go

```diff
@@ -145,48 +131,102 @@ func (r *DockerHubRegistry) loadBundleImageData(org string) ([]*ImageData, error
+	r.log.Debug("BundleSpecLabel: %s", BundleSpecLabel)
+	r.log.Debug("Loading image list for org: [ %s ]", org)
+
+	type Images struct {
+		Name      string `json:"name"`
+		Namespace string `json:"namespace"`
 	}
-	req.Header.Set("Authorization", os.ExpandEnv(fmt.Sprintf("JWT %v", string(tokenResp.Token))))
 
-	resp, err = http.DefaultClient.Do(req)
-	if err != nil {
-		return nil, err
+	type ImageResponse struct {
+		Count   int       `json:"count"`
+		Results []*Images `json:"results"`
+		Next    string    `json:"next"`
 	}
-	defer resp.Body.Close()
 
-	imageList, err := ioutil.ReadAll(resp.Body)
+	token, err := r.getDockerHubToken()
 
-	imageResp := ImageResponse{}
-	err = json.Unmarshal(imageList, &imageResp)
-	if err != nil {
-		return nil, err
+	channel := make(chan *ImageData)
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+
+	// Will kick of go routines for each result in the images.
+	loadImagesFromResult := func(ctx context.Context, images []*Images) {
+		for _, imageName := range images {
+			r.log.Debugf("Trying to load %v/%v", imageName.Namespace, imageName.Name)
+			go r.loadImageData(ctx, "docker://"+imageName.Namespace+"/"+imageName.Name, channel)
+		}
 	}
 
-	channel := make(chan *ImageData)
+	// getNextImages - will follow the next URL using go routines.
+	// The workflow is query the url, if Next is defined then kickoff another go routine to get those images.
+	// then kick of the loading of image data.
+	var getNextImages func(ctx context.Context, org, token, url string, ch chan<- *ImageData, cancelFunc context.CancelFunc) (*ImageResponse, error)
+
+	getNextImages = func(ctx context.Context, org, token, url string, ch chan<- *ImageData, cancelFunc context.CancelFunc) (*ImageResponse, error) {
+		req, err := http.NewRequest("GET", url, nil)
+		if err != nil {
+			r.log.Errorf("unable to get next images for url: %v - %v", url, err)
+			cancelFunc()
+			close(ch)
+			return nil, err
+		}
+		req.Header.Set("Authorization", fmt.Sprintf("JWT %v", token))
+
+		resp, err := http.DefaultClient.Do(req)
+		if err != nil {
+			r.log.Errorf("unable to get next images for url: %v - %v", url, err)
+			cancelFunc()
+			close(ch)
+			return nil, err
+		}
+		defer resp.Body.Close()
 
-	for _, imageName := range imageResp.Results {
-		r.log.Debug("Trying to load " + imageName.Namespace + "/" + imageName.Name)
-		go r.loadImageData("docker://"+imageName.Namespace+"/"+imageName.Name, channel)
+		imageList, err := ioutil.ReadAll(resp.Body)
+
+		iResp := ImageResponse{}
+		err = json.Unmarshal(imageList, &iResp)
+		if err != nil {
+			r.log.Errorf("unable to get next images for url: %v - %v", url, err)
+			cancelFunc()
+			close(ch)
+			return &iResp, err
+		}
+		//Keep getting the images
+		if iResp.Next != "" {
+			r.log.Debugf("getting next page of results - %v", iResp.Next)
+			//Fan out calls to get the next images.
+			go getNextImages(ctx, org, token, iResp.Next, channel, cancelFunc)
```

Comment:
I can't come up with a scenario to where the threads hang around. The channel closing looks good.

We may not be able to even use a waitgroup because I think we would need to know the number of threads to wait for. And it could be that when we run it on line 251, some of them have already completed.  I think this patch looks good. If we have a problem later we could always come back and fix it.